src/containers.c

```diff
@@ -0,0 +1,314 @@
+    uint32_t timeout = 0;
+    if (ConfGetValue("containers.memcap", &str) == 1) {
+        if (ParseSizeStringU64(str, &memcap) < 0) {
+            SCLogWarning(SC_ERR_INVALID_VALUE,
+                    "memcap value cannot be deduced: %s,"
+                    " resetting to default",
+                    str);
+            memcap = 0;
+        }
+    }
+    if (ConfGetValue("containers.timeout", &str) == 1) {
+        if (ParseSizeStringU32(str, &timeout) < 0) {
+            SCLogWarning(SC_ERR_INVALID_VALUE,
+                    "timeout value cannot be deduced: %s,"
+                    " resetting to default",
+                    str);
+            timeout = 0;
+        }
+    }
+
+    ContainerUrlRangeList.ht = THashInit("containers.urlrange", sizeof(ContainerUrlRange),
+            ContainerUrlRangeSet, ContainerUrlRangeFree, ContainerUrlRangeHash,
+            ContainerUrlRangeCompare, false, memcap, CONTAINER_URLRANGE_HASH_SIZE);
+    ContainerUrlRangeList.timeout = timeout;
+
+    SCLogDebug("containers started");
+}
+
+void ContainersDestroy(void)
+{
+    THashShutdown(ContainerUrlRangeList.ht);
+}
+
+uint32_t ContainersTimeoutHash(struct timeval *ts)
+{
+    uint32_t cnt = 0;
+
+    for (size_t i = 0; i < ContainerUrlRangeList.ht->config.hash_size; i++) {
+        THashHashRow *hb = &ContainerUrlRangeList.ht->array[i];
+
+        if (HRLOCK_TRYLOCK(hb) != 0)
+            continue;
+        /* hash bucket is now locked */
+        THashData *h = hb->head;
+        while (h) {
+            THashData *n = h->next;
+            if (ContainerValueRangeTimeout(h->data, ts)) {
+                /* remove from the hash */
+                if (h->prev != NULL)
+                    h->prev->next = h->next;
+                if (h->next != NULL)
+                    h->next->prev = h->prev;
+                if (hb->head == h)
+                    hb->head = h->next;
+                if (hb->tail == h)
+                    hb->tail = h->prev;
+                h->next = NULL;
+                h->prev = NULL;
+                ContainerUrlRange *c = h->data;
+                FileCloseFile(c->files, NULL, 0, FILE_TRUNCATED);
+                // TODOask can we log it somehow ?
+                RangeContainerFree(c);
+                THashDataMoveToSpare(ContainerUrlRangeList.ht, h->data);
+            }
+            h = n;
+        }
+        HRLOCK_UNLOCK(hb);
+    }
+
+    return cnt;
+}
+
+void *ContainerUrlRangeGet(const uint8_t *key, size_t keylen, struct timeval *ts)
+{
+    ContainerUrlRange lookup;
+    // cast so as not to have const in the structure
+    lookup.key = (uint8_t *)key;
+    lookup.len = keylen;
+    struct THashDataGetResult res = THashGetFromHash(ContainerUrlRangeList.ht, &lookup);
+    if (res.data) {
+        // nothing more to do if (res.is_new)
+        ContainerUrlRangeUpdate(res.data->data, ts->tv_sec + ContainerUrlRangeList.timeout);
+        THashDecrUsecnt(res.data);
+        THashDataUnlock(res.data);
```

Comment:
data is now no long protected by lock, but caller gets a reference to it and proceeds to do things with it. This seems to be a thread safety issue.

Reply:
You are right.
There are 2 problems : 
- the fact that `ContainerUrlRange` has no inner lock/mutex and that multiple flows may use it (read/write)
Solution is to add a lock to the structure and use it
- The fact that the item of the hash table is saved in a flow structure cf `file_range_container` and that it may timeout and get removed/freed of the hash table while remaining in the flow structure
Solution may be to add a number of references to the structure

Reply:
Done the easy part about the number of references

Now remains the hard part with multiple flows writing concurrently to the file's ranges
It looks like `ContainerUrlRange` needs to be split in 2 :
- one in the hash table where I should open and close a range
- one lockless structure per flow for the current range state (ie fields `toskip`, `current`) when appending data